src/pluto_eris/engine.rs

```diff
@@ -0,0 +1,942 @@
+        fn exp_by_x(f: &mut Fp12) {
+            let x = NEG_PLUTO_U;
+            let mut res = Fp12::ONE;
+            for i in (0..111).rev() {
+                res.cyclotomic_square();
+                if ((x >> i) & 1) == 1 {
+                    res.mul_assign(f);
+                }
+            }
+            res.conjugate();
+            *f = res;
+        }
+
+        let r = self.0;
+        let mut f1 = self.0;
+        f1.conjugate();
+
+        Gt(r.invert()
+            .map(|mut f2| {
+                let mut r = f1;
+                r.mul_assign(&f2);
+                f2 = r;
+                r.frobenius_map(2);
+                r.mul_assign(&f2);
+
+                let mut fp = r;
+                fp.frobenius_map(1);
+
+                let mut fp2 = r;
+                fp2.frobenius_map(2);
+                let mut fp3 = fp2;
+                fp3.frobenius_map(1);
+
+                let mut fu = r;
+                exp_by_x(&mut fu);
+
+                let mut fu2 = fu;
+                exp_by_x(&mut fu2);
+
+                let mut fu3 = fu2;
+                exp_by_x(&mut fu3);
+
+                let mut y3 = fu;
+                y3.frobenius_map(1);
+
+                let mut fu2p = fu2;
+                fu2p.frobenius_map(1);
+
+                let mut fu3p = fu3;
+                fu3p.frobenius_map(1);
+
+                let mut y2 = fu2;
+                y2.frobenius_map(2);
+
+                let mut y0 = fp;
+                y0.mul_assign(&fp2);
+                y0.mul_assign(&fp3);
+
+                let mut y1 = r;
+                y1.conjugate();
+
+                let mut y5 = fu2;
+                y5.conjugate();
+
+                y3.conjugate();
+
+                let mut y4 = fu;
+                y4.mul_assign(&fu2p);
+                y4.conjugate();
+
+                let mut y6 = fu3;
+                y6.mul_assign(&fu3p);
+                y6.conjugate();
+
+                y6.cyclotomic_square();
+                y6.mul_assign(&y4);
+                y6.mul_assign(&y5);
+
+                let mut t1 = y3;
+                t1.mul_assign(&y5);
+                t1.mul_assign(&y6);
+
+                y6.mul_assign(&y2);
+
+                t1.cyclotomic_square();
+                t1.mul_assign(&y6);
+                t1.cyclotomic_square();
+
+                let mut t0 = t1;
+                t0.mul_assign(&y1);
+
+                t1.mul_assign(&y0);
+
+                t0.cyclotomic_square();
+                t0.mul_assign(&t1);
+
+                t0
+            })
+            .unwrap())
+    }
+}
+
+pub fn multi_miller_loop(terms: &[(&G1Affine, &G2Prepared)]) -> Gt {
+    let mut pairs = vec![];
+    for &(p, q) in terms {
+        if !bool::from(p.is_identity()) && !q.is_zero() {
+            pairs.push((p, q.coeffs.iter()));
+        }
+    }
+
+    // Final steps of the line function on prepared coefficients
+    fn ell(f: &mut Fp12, coeffs: &(Fp2, Fp2, Fp2), p: &G1Affine) {
```

Comment:
I think this could be a good improvement!
I have been reading a bit about autovectorisation but it is unclear to me what are the conditions that need to be met for it to apply. 

This will be a good extra improvement for the refactor in #76.

Reply:
It's usually hard to tell as depends only on the compiler. But I know at least in the past, Iterators were guaranteed to take profit of it. Not for loops where it wasn't always the case though.

In any case, makes sense to address later!